presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java

```diff
@@ -745,39 +757,81 @@ public ConnectorPageSource createPageSource(
                 .map(IcebergColumnHandle.class::cast)
                 .collect(toImmutableList());
 
-        Map<Integer, HivePartitionKey> partitionKeys = split.getPartitionKeys();
-
-        List<IcebergColumnHandle> regularColumns = columns.stream()
-                .map(IcebergColumnHandle.class::cast)
-                .filter(column -> column.getColumnType() != PARTITION_KEY &&
-                        !partitionKeys.containsKey(column.getId()) &&
-                        !IcebergMetadataColumn.isMetadataColumnId(column.getId()))
-                .collect(Collectors.toList());
-
         Optional<String> tableSchemaJson = table.getTableSchemaJson();
         verify(tableSchemaJson.isPresent(), "tableSchemaJson is null");
         Schema tableSchema = SchemaParser.fromJson(tableSchemaJson.get());
+        PartitionSpec partitionSpec = PartitionSpecParser.fromJson(tableSchema, split.getPartitionSpecAsJson());
 
-        boolean equalityDeletesRequired = table.getIcebergTableName().getTableType() == IcebergTableType.DATA;
-        Set<IcebergColumnHandle> deleteFilterRequiredColumns = requiredColumnsForDeletes(tableSchema, split.getDeletes(), equalityDeletesRequired);
+        Map<Integer, HivePartitionKey> partitionKeys = split.getPartitionKeys();
+
+        // the update row isn't a valid column that can be read from storage.
+        // Filter it out from columns passed to the storage page source.
+        Set<IcebergColumnHandle> columnsToReadFromStorage = icebergColumns.stream()
+                .filter(column -> UPDATE_ROW_DATA.getId() != column.getId())
+                .collect(Collectors.toSet());
 
-        deleteFilterRequiredColumns.stream()
+        // add any additional columns which may need to be read from storage
+        // by delete filters
+        boolean equalityDeletesRequired = table.getIcebergTableName().getTableType() == IcebergTableType.DATA;
+        requiredColumnsForDeletes(tableSchema, partitionSpec, split.getDeletes(), equalityDeletesRequired)
+                .stream()
                 .filter(not(icebergColumns::contains))
-                .forEach(regularColumns::add);
+                .forEach(columnsToReadFromStorage::add);
+
+        // finally, add the fields that the update column requires.
+        Optional<IcebergColumnHandle> updateRow = icebergColumns.stream()
+                .filter(IcebergColumnHandle::isUpdateRowIdColumn)
+                .findFirst();
+        updateRow.ifPresent(updateRowIdColumn -> {
+            Set<Integer> alreadyRequiredColumnIds = columnsToReadFromStorage.stream()
+                    .map(IcebergColumnHandle::getId)
+                    .collect(toImmutableSet());
+            updateRowIdColumn.getColumnIdentity().getChildren()
```

Comment:
nit: we can probably store it in a variable before streaming 
`List<ColumnIdentity> = updateRowIdColumn.getColumnIdentity().getChildren();`

Reply:
What's the benefit of storing the variable? I don't see where it would be used later

Reply:
I agree I also don't see any usage, I probably misread something due to which I wrote this comment. We can ignore this.